feat(typing): Add type hinting to the trace view

wmak opened the pull request:
- This adds type hints to the trace view
- As far as I could tell, py3.6 doesn't support typedicts yet, but to
  keep it easier for later I've added TraceError and SnubaEvent at the
  top that can be later changed to TypeDicts
- Moved the event that is passed around to a class for a similar reason,
  but also refactored how its used between the two endpoints too

wmak committed:
feat(typing): Add type hinting to the trace view

- This adds type hints to the trace view
- As far as I could tell, py3.6 doesn't support typedicts yet, but to
  keep it easier for later I've added TraceError and SnubaEvent at the
  top that can be later changed to TypeDicts
- Moved the event that is passed around to a class for a similar reason,
  but also refactored how its used between the two endpoints too

untitaker commented on src/sentry/api/endpoints/organization_events_trace.py:
```diff
@@ -96,30 +162,15 @@ def query_trace_data(trace_id, params):
-    def serialize_error(self, event):
+    @staticmethod
+    def serialize_error(event: SnubaEvent) -> TraceError:
         return {
```
TraceError says all of its values are strings. is project_id really a string? may consider TypedDict

evanh replied:
Similarly "parent_span_id" is actually an `Optional[str]`. and "errors" is a list. So the typing doesn't hold up here.

untitaker commented on src/sentry/api/endpoints/organization_events_trace.py:
```diff
@@ -17,20 +20,83 @@
 
+SnubaEvent = Mapping[str, str]
+TraceError = Mapping[str, str]
+ResponseType = Dict[str, Any]
```
Consider TypedDict

untitaker commented on src/sentry/api/endpoints/organization_events_trace.py:
```diff
@@ -17,20 +20,83 @@
+ResponseType = Dict[str, Any]
+
+
+class TraceEvent:
```
this class doesn't do a whole lot of stuff either, may also consider typeddict here but it's not like this is any worse. just keeps the diff smaller

untitaker commented:
@wmak sorry overlooked your comment on TypedDict. You can do:

```
if False:
    from mypy_extensions import TypedDict
    
    SomeType = TypedDict(...)
    
def foo(x: 'SomeType') -> None: ...
```

the alternative is to add mypy_extensions to production requirements lol

wmak commented:
😄 @untitaker no worries, I also thought I could use TypedDict at first.

evanh commented on src/sentry/api/endpoints/organization_events_trace.py:
```diff
@@ -17,20 +20,83 @@
+
 
-def find_event(items, function, default=None):
+# Can't do Callable[[Any], bool] for the function cause anything on an Any -> Any
```
Can you elaborate on this?

wmak replied:
Hmm actually looking again at this, I think my comment is wrong and I can type this
```python
def find_event(items: Sequence[Mapping[str, str]], function: Callable[[Mapping[str, str]], bool], default: Optional[Mapping[str, str]] = None) -> Optional[Mapping[str, str]]:
```

But now its complaining about https://github.com/getsentry/sentry/pull/25432/files#diff-619b86f549a9f11a566ca757be8da43a63bcaf6e4d7d3fe88a4222937a0209c6R274 being not indexed 🤔 I'll keep trying

evanh commented on src/sentry/api/endpoints/organization_events_trace.py:
```diff
@@ -96,30 +162,15 @@ def query_trace_data(trace_id, params):
-    def has_feature(self, organization, request):
+# TODO(wmak): remove ignore once the base is typed
+class OrganizationEventsTraceEndpointBase(OrganizationEventsV2EndpointBase):  # type: ignore
+    def has_feature(self, organization: Organization, request: HttpRequest) -> Any:
```
This should return a bool right? If `features.has` returns any you should do `assert isinstance(x, bool)` or `bool(x)`.

evanh reviewed (commented):
If you are looking for an alternative to TypedDict, I would use dataclasses. They are a third party package for 3.6 (you'd need to add it to the requirements) but otherwise is exactly the same as the built in dataclass once we upgrade to 3.8. If you wanted to add that package to Sentry it would probably help out a lot of other people who are going to run into a similar situation to yourself.

wmak committed:
ref: Addressing the easier comments

- Casting feature, since its currently untyped
- typing find_event

wmak committed:
ref: Trying to use TypedDict

- basically make TypedDict a small no-op thing in prod, but when type
  checking use the mypy_extensions.TypedDict since we'll have it
- looked into dataclasses, but they would've required some extra work to
  get serialization working

wmak commented:
@evanh @untitaker 
So I looked into Dataclasses, but it looked like there would have to be extra work to get serialization working on them so I ended up trying to get TypedDicts to work without requiring an extra requirement in prod.
`typing.TYPE_CHECKING` should be False everywhere but when mypy is type checking so I think https://github.com/getsentry/sentry/blob/2fefd6dee8900693d9f381b6436d871758fab9d0/src/sentry/api/endpoints/organization_events_trace.py#L36-L41

should be a decent workaround until a later version of python

wmak commented on src/sentry/api/endpoints/organization_events_trace.py:
```diff
@@ -1,36 +1,211 @@
+        "event_id": str,
+        "span_id": str,
+        "transaction": str,
+        "transaction.duration": int,
```
Is there a way to get keys with periods to work with the class based TypedDict? because the FullResponse is identical to the Light Response except for a few extra keys.

evanh replied:
What's the problem with keys with periods?

untitaker replied:
a class-based typeddict uses python identifiers to define keys, and `.` is not valid in identifiers

evanh replied:
Ah, thank you. @wmak They work in 3.8, so maybe it has to stay this way for now. One thing that _might_ work, is if all the keys that were in FullResponse that are NOT part of LightResponse were valid identifiers, you could write LightResponse using the function method and then use FullResponse extending LightResponse. I'm not sure that would work though.

evanh commented on src/sentry/api/endpoints/organization_events_trace.py:
```diff
@@ -164,10 +167,11 @@ def query_trace_data(trace_id: str, params: Mapping[str, str]) -> Sequence[Seque
 
 # TODO(wmak): remove ignore once the base is typed
```
You don't actually need these, mypy will raise an error if you do `type: ignore` on a line that doesn't have an error. So if someone does type the base they'll have to fix this in order to get the typing to pass.

wmak replied:
woah, that's awesome!

evanh reviewed (approved):
This looks good to me. Well done!

Can you elaborate on the issue with serializing dataclasses?

wmak commented:
@evanh basically that I can't just swap the dataclaass in and still treat the object like a dict eg.
```python
from dataclasses import dataclass

@dataclass
class Test:
    a :str
test = Test(a="foo")

json.dumps(test)
# TypeError: Object of type 'Test' is not JSON serializable
test["a"] == "bar"
# TypeError: 'Test' object does not support item assignment
```

I could update the code to treat everything that ended up as a typeDict as a class instead, but I didn't want to change the endpoint that significantly

wmak committed:
ref: Removing unneccesary TODO

- mypy will raise an error if type: ignore is on a line that doesn't
  have an error

evanh commented:
> I could update the code to treat everything that ended up as a typeDict as a class instead, but I didn't want to change the endpoint that significantly

Got it, that makes total sense.

wmak committed:
ref: Add a comment explaining the TypeDict hack

wmak committed:
Merge branch 'master' into wmak/feat/type-trace-endpoint

wmak committed:
ref: Just ignore rest_framework instead of having to ignore on import
